tests/lib/prepare.sh

```diff
@@ -850,17 +850,67 @@ EOF
         # need to download it
         snap download --channel="$KERNEL_CHANNEL" pc-kernel
 
-        EXTRA_FUNDAMENTAL="--extra-snaps $PWD/pc-kernel_*.snap"
+        EXTRA_FUNDAMENTAL="--snap $PWD/pc-kernel_*.snap"
         IMAGE_CHANNEL="$GADGET_CHANNEL"
     fi
 
     if os.query is-core20; then
+        # on UC20 we always need to repack the kernel to use our version of 
+        # snap-bootstrap and friends in the initrd so do that now
         snap download --basename=pc-kernel --channel="20/$KERNEL_CHANNEL" pc-kernel
         # make sure we have the snap
         test -e pc-kernel.snap
         # build the initramfs with our snapd assets into the kernel snap
         uc20_build_initramfs_kernel_snap "$PWD/pc-kernel.snap" "$IMAGE_HOME"
-        EXTRA_FUNDAMENTAL="--extra-snaps $IMAGE_HOME/pc-kernel_*.snap"
+        EXTRA_FUNDAMENTAL="--snap $IMAGE_HOME/pc-kernel_*.snap"
+
+        # also add debug command line parameters to the kernel command line via
+        # the gadget in case things go side ways and we need to debug
+        snap download --basename=pc --channel="20/$GADGET_CHANNEL" pc
+        test -e pc.snap
+        unsquashfs -d pc-gadget pc.snap
+        
+        # TODO: it would be desirable when we need to do in-depth debugging of
+        # UC20 runs in google to have snapd.debug=1 always on the kernel command
+        # line, but we can't do this universally because the logic for the env
+        # variable SNAPD_DEBUG=0|false does not overwrite the turning on of 
+        # debug messages in some places when the kernel command line is set, so
+        # we get failing tests since there is extra stuff on stderr than 
+        # expected in the test when SNAPD_DEBUG is turned off
+        # so for now, don't include snapd.debug=1, but eventually it would be
+        # nice to have this on
+
+        if [ "$SPREAD_BACKEND" = "google" ]; then
+            # the default console settings for snapd aren't super useful in GCE,
+            # instead it's more useful to have all console go to ttyS0 which we 
+            # can read more easily than tty1 for example
+            for cmd in "console=ttyS0" "dangerous" "systemd.journald.forward_to_console=1" "rd.systemd.journald.forward_to_console=1" "panic=-1"; do
+                echo "$cmd" >> pc-gadget/cmdline.full
```

Comment:
Do we have some checks somewhere to ensure that we are not hitting the maximum kernel command length? Just asking because I've hit that problem before when playing with an Android-based phone (there the max was 512, IIRC) :-)

If we are way below 512 characters no worries, though :-)

Reply:
Hmm I don't think we have such a validation, but I'll leave a TODO